feat(elements): create custom elements without NgModule

pkozlowski-opensource opened the pull request:
This is an exploratory PR to introduce APIs needed for Angular elements creation without NgModule. It has 2 changes (will split it into several commits before sending out for review):
* introduces the notation of an environment and the `createRootEnvironment` API;
* tests the `createRootEnvironment` API in the context of custom elements created from standalone components;

There are several open questions on the table but opening an early draft PR to discuss details of the APIs.

michielkikkert commented:
Very much looking forward to this one, thanks! I’m curious about the ergonomics of this implementation.

pkozlowski-opensource commented:
@michielkikkert you can get the feel for the ergonomics by looking at the [test](https://github.com/angular/angular/pull/46475/files#diff-3d645d47eeac056c97e20b4457b131329467e6a95e0ce7de78e5048e87bb6db7R29-R51).

But the tl;dr; is this:

```typescript
// create a constructor of a custom element
const NgElementCtor = createCustomElement(TestStandaloneCmp, {injector: envInjector});

// register in a browser
customElements.define('test-standalone-cmp', NgElementCtor);

// good to go!
```

manfredsteyer commented:
I'm really happy to see this 🧡

Is there an option to bootstrap the custom element with noop-zone or is this subject to another RFC/ topic/ PR?

pkozlowski-opensource committed:
feat(core): introduce createApplication API

The `createApplication` function makes it possible to create an
application instance (represented by the `ApplicationRef`)
without bootstrapping any components. It is useful in the
situations where ones wants to decouple and delay components
rendering and / or render multiple root components in one
application. Angular elements can use this API to create
custom element types with an environment linked to a
created application.

pkozlowski-opensource committed:
test(elements): add a test for standalone element creation

This commit adds a test demonstrating how a standalone Angular component
could be turned into a custom element.

pkozlowski-opensource commented on packages/elements/test/create-custom-element-env_spec.ts:
```diff
@@ -0,0 +1,53 @@
+      const appRef = await createRootEnvironment();
+
+      try {
+        const NgElementCtor = createCustomElement(TestStandaloneCmp, {injector: appRef.injector});
```
With the approach tested here we don't enforce that `TestStandaloneCmp` is a standalone one. This has a sharp edge but doesn't require API modification.

An alternative approach would be to add a new option field - `environmentInjector` - to the `createCustomElement` and enforce `standalone` bit when this new option is used.

@gkalpak @AndrewKushnir @alxhub any thoughts?

gkalpak replied:
I might be missing something, but here is my thinking:

We shouldn't care if the component is standalone or not. It is already possible to create Angular Elements with standalone components. This PR is more about creating them with a root environment injector; i.e. without requiring an `NgModule`.

If my thinking is correct, we should not care if (and therefore not enforce that) the component is standalone.
(I would even update the commit message to reflect that - i.e. to put emphasis on "without an `NgModule`" rather than "with a standalone component".)

pkozlowski-opensource replied:
> If my thinking is correct, we should not care if (and therefore not enforce that) the component is standalone.

I agree that we could see it this way. It is just that, today, using this API with non-standalone components is "inherently dangerous" - I can easily specify a component that doesn't belong to a NgModule, belongs to a different module etc. Such component would, most probably, fail at runtime as they would lack NgModule context (ex. collaborating directives, providers etc.)

So I hear what you are saying, and I agree that if we are leaving things as-is we are not making the situation worse. But maybe there is an opportunity for improvement here?

gkalpak replied:
Oh, I see your point. That's a good question.

I feel that the extra mental overhead of introducing a separate way for using this API with standalone components is not worth the benefit we get. If I am not mistaken, you could still run into the same pitfalls (for example, by using the current/"old" way of registering the Angular element).

I don't feel strongly about it (and it is also possible I am missing some internal details regarding standalone components), but I would vote for keeping things as they are for now and revisit this once the dust around standalone APIs (and the associated usecases/needs) has settled :smiley:

pkozlowski-opensource replied:
Yeh. Let's leave it as-is for now.

gkalpak commented on packages/core/src/application_ref.ts:
```diff
@@ -223,27 +227,30 @@ export function internalBootstrapApplication(config: {
-        setLocaleId(localeId || DEFAULT_LOCALE_ID);
 
-        const appRef = appInjector.get(ApplicationRef);
+    envInjector.onDestroy(() => onErrorSubscription.unsubscribe());
```
Would it make sense to merge this with the other `onDestroy()` listener (on line 238)?

pkozlowski-opensource replied:
Yep, sgtm, thnx!

gkalpak commented on packages/core/src/application_ref.ts:
```diff
@@ -223,27 +227,30 @@ export function internalBootstrapApplication(config: {
+      initStatus.runInitializers();
 
-        appRef.bootstrap(rootComponent);
+      const localeId = envInjector.get(LOCALE_ID, DEFAULT_LOCALE_ID);
+      setLocaleId(localeId || DEFAULT_LOCALE_ID);
```
Is it possible that something run during the app initialization would change the `LOCALE_ID`?

pkozlowski-opensource replied:
Not sure, TBH. Are you thinking about any specific timing issue here?

gkalpak replied:
Previously, the `envInjector.get(LOCALE_ID, DEFAULT_LOCALE_ID)` was inside the `initStatus.donePromise` callback. AFAICT, this means that if one of the async app initializers did something that affected the value of `LOCALE_ID`, then it matters whether the token is requested from the injector before or after `initStatus.donePromise` is fulfilled.

For example, imagine something like the following:

```
let localeId = 'en';  // Default locale.

@NgModule({
  ...
  providers: [
    { provide: LOCALE_ID, useFactory: () => localeId },
    {
      provide APP_INITIALIZER,
      multi: true,
      useValue: async () => {
        const response = await fetch('/user/preferences/locale');
        localeId = await response.json();
      },
    },
  ],
})
```

pkozlowski-opensource replied:
Yeh, I hear you. No strong opinion here so moving it back to where it was.

gkalpak commented on packages/platform-browser/src/browser.ts:
```diff
@@ -96,14 +96,34 @@ export interface ApplicationConfig {
+ * associated injectors) from rendering components on a screen. Components can be subsequently
+ * bootstrapped on the returned `ApplicationRef`.
+ *
+ * @param options Extra configuration for the application environnement, see `ApplicationConfig` for
```
```suggestion
 * @param options Extra configuration for the application environment, see `ApplicationConfig` for
```

pkozlowski-opensource replied:
👍

gkalpak commented on packages/platform-browser/src/browser.ts:
```diff
@@ -96,14 +96,34 @@ export interface ApplicationConfig {
+ * @publicApi
+ * @developerPreview
+ */
+export function createRootEnvironment(options?: ApplicationConfig) {
```
I find it a little confusing that we have this function named `createRootEnvironment` which returns an `ApplicationRef`. It sounds like we have now two different terms (`RootEnvironment` and `Application`/`ApplicationRef`) refer to the same thing.

Would it make sense to unify them somehow?

pkozlowski-opensource replied:
Agreed. This is still being debated.

pkozlowski-opensource replied:
Settling on `createApplication` based on all the feedback / discussions so far.

gkalpak commented on packages/elements/test/create-custom-element-env_spec.ts:
```diff
@@ -0,0 +1,53 @@
+      @Component({
+        standalone: true,
+        template: `Hello, standalone element!`,
+
```
Nit:
```suggestion
```

gkalpak reviewed (commented):
So looking forward to this :tada:
Thank you, @pkozlowski-opensource 👏

I left a couple of minor comments/questions, but overall lgtm :+1:

BTW, there are a couple of typos in the second commit message:
- `where ones want to` --> `where one wants to`
- `environnement` --> `environment`

pkozlowski-opensource commented:
@gkalpak thnx for the detailed review and all the feedback. I believe that I've addressed most of it, feel free to do another pass on this PR.

AndrewKushnir commented on goldens/public-api/platform-browser/index.md:
```diff
@@ -62,6 +62,9 @@ export class By {
 }
 
+// @public
+export function createApplication(options?: ApplicationConfig): Promise<ApplicationRef>;
```
👍

AndrewKushnir commented on packages/core/src/application_ref.ts:
```diff
@@ -187,16 +187,19 @@ export function runPlatformInitializers(injector: Injector): void {
  * @returns A promise that returns an `ApplicationRef` instance once resolved.
  */
 export function internalBootstrapApplication(config: {
```
Just as a thought here: what if we:
- rename this function to `internalCreateApplication`
- drop the `rootComponent` logic from it
- introduce a new `internalBootstrapApplication` that would use `internalCreateApplication` under the hood
- use corresponding functions in the `createApplication` and `bootstrapApplication` public functions

It feels a bit backwards that the `createApplication` uses `internalBootstrapApplication` internally (it feels more organic to create an app and call bootstrap after that). The `internalBootstrapApplication` may look like this:

```
export function internalBootstrapApplication(config: {...}): Promise<ApplicationRef> {
  const {rootComponent, appProviders, platformProviders} = config;
  NG_DEV_MODE && assertStandaloneComponentType(rootComponent);

  const bootstrapComponent = (appRef: ApplicationRef) => appRef.bootstrap(rootComponent);
  return internalCreateApplication({...config, callback: bootstrapComponent});
}
```

(the `callback` is needed, so that we can invoke the function within the `_callAndReportToErrorHandler` callback fn, which is invoked in try-catch)

Just an idea, feel free to ignore :) We can do it later if needed, since it's all internal implementation.

pkozlowski-opensource replied:
I hear you. Actually I had a version with a callback but found it to be a bit of "too much".

I've renamed `internalBootstrapApplication` to `internalCreateApplication` and kept the `rootComponent` as an optional parameter. IMO this is good for the first iteration, I wouldn't like to block this PR any more but happy to discuss further refactoring in a separate PR.

AndrewKushnir reviewed (approved):
@pkozlowski-opensource the API looks really nice 👍 

Reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: fw-core, public-api, fw-platform-server

dylhunn reviewed (approved):
reviewed-for: public-api, fw-core

jessicajaniuk commented:
This PR was merged into the repository by commit 9285c6613eec5138037f72b4036f58e281c4742b.
